aii-pxelinux/src/main/perl/pxelinux.pm

```diff
@@ -493,33 +640,69 @@ sub Unconfigure
-    return 1;
+    my $interfaces = $cfg->getElement (INTERFACES)->getTree;
+
+    foreach my $variant_constant (@PXE_VARIANTS) {
+        my $variant = __PACKAGE__->$variant_constant;
+        my $pxe_config_file = $self->_filepath ($cfg, $variant);
+        # Remove the PXEe config file for the current host
+        $self->debug(1, "Removing PXE config file $pxe_config_file (PXE variant=",
+                            $self->_variant_attribute('name', $variant), ")");
+        my $unlink_status = $self->cleanup($pxe_config_file);
+        if ( ! defined($unlink_status) ) {
+            $self->error("Failed to delete $pxe_config_file (error=$self->{fail})");
+        } elsif ( $unlink_status == SUCCESS ) {
+            $self->debug(1, "PXE config file $pxe_config_file not found");
+        } else {
+            $self->debug(1, "PXE config file $pxe_config_file successfully removed");
+        };
+        # Remove the symlink for every interface with an IP address
+        while (my ($interface, $params) = each %$interfaces) {
+            if ( defined($params->{ip}) ) {
+                my $pxe_symlink =  dirname($pxe_config_file) . "/" . $self->_hexip_filename ($params->{ip}, $variant);
+                $self->debug(1, "Removing symlink $pxe_symlink for interface $interface (PXE variant=",
+                                    $self->_variant_attribute('name', $variant), ")");
+                my $unlink_status = $self->cleanup($pxe_symlink);
+                if ( ! defined($unlink_status) ) {
+                    $self->error("Failed to delete $pxe_symlink (error=$self->{fail})");
+                } elsif ( $unlink_status == SUCCESS ) {
+                    $self->debug(1, "PXE link $pxe_symlink not found");
+                } else {
+                    $self->debug(1, "PXE link $pxe_symlink successfully removed");
+                };
+            };
+        };
+        $self->_exec_userhooks ($cfg, REMOVE_HOOK_PATH);
+    }
+
+    return 1;        
 }
 
 
 no strict 'refs';
-foreach my $i (qw(configure boot rescue livecd firmware install)) {
-    my $name = ucfirst($i);
-    my $cmd = uc($i);
+foreach my $operation (qw(configure boot rescue livecd firmware install)) {
+    my $name = ucfirst($operation);
+    my $cmd = uc($operation);
 
     *{$name} = sub {
         my ($self, $cfg) = @_;
 
-        pxeprint($cfg) if ($i eq 'configure');
-
-        unless (pxelink ($cfg, &$cmd())==0) {
-            my $fqdn = get_fqdn($cfg);
-            $self->error ("Failed to change the status of $fqdn to $i");
-            return 0;
+        foreach my $variant_constant (@PXE_VARIANTS) {
+            my $variant = __PACKAGE__->$variant_constant;
+            if ( $self->_variant_enabled($variant) ) {
+                $self->verbose("Executing action '$operation' for variant ", $self->_variant_attribute('name', $variant));
+                $self->_variant_attribute('format_method', $variant)->($cfg) if ($operation eq 'configure');
```

Comment:
Same as https://github.com/quattor/aii/pull/225#pullrequestreview-24734356. A side effect of the previous problem as I was relying on the non-existent unit test to detect if the calling sequence was correct...